src/Composer/Package/Loader/RootPackageLoader.php

```diff
@@ -93,6 +99,15 @@ public function load(array $config, string $class = 'Composer\Package\RootPackag
             }
 
             if (!isset($config['version'])) {
+                if ($this->io !== null) {
```

Comment:
If no name is set, circular dependencies are impossible so warning about this is probably irrelevant - and as a nice side effect it'll fix most tests I think.

```suggestion
                if ($this->io !== null && $config['name'] !== '__root__') {
```